### PR TITLE
Adding dapt curation tutorial (incorporated feedback)

### PR DESCRIPTION
**Pull request by jvamaraju**
## Usage

```python
# Add snippet demonstrating usage
```
## Checklist

- [ ] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [ ] New or Existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: tutorials/dapt-curation/code/main.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/main.py:223]: [nit] Remove?

File: tutorials/dapt-curation/code/main.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/main.py:57]: [nit] Remove? I saw the TODO section inside README.md, so maybe we can just add the instructions there and remove these?

File: tutorials/dapt-curation/code/downloaders.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/downloaders.py:74]: [nit] Remove?

File: tutorials/dapt-curation/code/downloaders.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/downloaders.py:77]: [nit] Remove?

File: tutorials/dapt-curation/code/downloaders.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/downloaders.py:251]: We might want to remove this as well, since we were just using it for our own testing.

File: tutorials/dapt-curation/code/modifiers.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/modifiers.py:20]: Looks like this is duplicated? There is another version inside `utils.py`. Let's delete this file, and keep the other file, given that all the other filters are also in that file.

File: tutorials/dapt-curation/code/utils.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/utils.py:224]: I'll let the engineers comment whether there is a better way to do this.

**Review by Maghoumi — APPROVED** [review]
Thanks again Janaki for your contribution! LGTM

File: tutorials/dapt-curation/code/main.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/main.py:223]: Done

File: tutorials/dapt-curation/code/main.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/main.py:57]: done

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:74]: done

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:77]: done

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:251]: done

File: tutorials/dapt-curation/code/modifiers.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/modifiers.py:20]: done

File: tutorials/dapt-curation/code/utils.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/utils.py:224]: @ryantwolf Recommended this way. Can he review?

File: tutorials/dapt-curation/code/main.py
Comment (ayushdg) [inline tutorials/dapt-curation/code/main.py:45]: Nit these should be importable from the top level directly.
```
from nemo_curator import ExactDuplicates, Modify, ScoreFilter, Sequential
```

File: tutorials/dapt-curation/code/utils.py
Comment (ayushdg) [inline tutorials/dapt-curation/code/utils.py:42]: Same as above

File: tutorials/dapt-curation/code/utils.py
Comment (ayushdg) [inline tutorials/dapt-curation/code/utils.py:232]: Question: If we specify device="gpu" here will this work on cpu only systems?

**Review by ayushdg — APPROVED** [review]
Minor nit's/ comments but overall looks good to me! Thank you so much for this great tutorial that's very comprehensive

File: tutorials/dapt-curation/code/utils.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/utils.py:232]: Yes this still works only on cpu. looks like its defaulting to cpu if not available.

File: tutorials/dapt-curation/code/main.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/main.py:45]: done

File: tutorials/dapt-curation/code/utils.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/utils.py:42]: done, changed

**Comment by jvamaraju** [comment]
Thank you @ayushdg @Maghoumi. @ryantwolf  Can you help review?

**Review by ayushdg — APPROVED** [review]
Thanks!

**Comment by ryantwolf** [comment]
I'll take a look at this today.

File: tutorials/dapt-curation/code/downloaders.py
Comment (ryantwolf) [inline tutorials/dapt-curation/code/downloaders.py:213]: Why is this the only set of urls that is being printed?

File: tutorials/dapt-curation/code/utils.py
Comment (ryantwolf) [inline tutorials/dapt-curation/code/utils.py:293]: Could you please use the `nemo_curator.Filter` module for this?

File: tutorials/dapt-curation/code/utils.py
Comment (ryantwolf) [inline tutorials/dapt-curation/code/utils.py:304]: Could you please use the `nemo_curator.Filter` module for this?

**Review by ryantwolf — CHANGES_REQUESTED** [review]
Overall looks good, thanks for making this. My major points just have to do with the downloading portion and turning `filter_text_lines` and `filter_code_lines` into something that can be used with `nemo_curator.Filter`.

File: tutorials/dapt-curation/code/docbuilder.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/docbuilder.py:215]: Let's remove `.vx` and `.vxh`  as those are proprietary extensions that won't exist in the GitHub repos.

File: tutorials/dapt-curation/code/main.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/main.py:120]: (Comment applies to the block below as well)

I think the ordering needs to change a bit:

1. clean and unify
2. dedupe
3. filter

The reason for doing `unify` first is to ensure dedupe operates on unified text and finds (potentially) more duplicates.

File: tutorials/dapt-curation/code/main.py
Comment (Maghoumi) [inline tutorials/dapt-curation/code/main.py:132]: Same as above.

File: tutorials/dapt-curation/code/utils.py
Comment (Maghoumi) [reply tutorials/dapt-curation/code/utils.py:293]: This can be two simple `ScoreFilter`s, no? 
- A ScoreFilter that returns `False` if the file type is not `text`.
- A ScoreFilter that returns `False` if lines < 10.

File: tutorials/dapt-curation/.gitignore
Comment (jvamaraju) [reply tutorials/dapt-curation/.gitignore:1]: @ryantwolf I have a jupyter notebook that I use to just test the code. Once all the changes have been approved and ok to merge I can delete this one.

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:7]: done

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:5]: done

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:3]: I meant using Llama 2 NIMs in the nemo format. Updated to reflect NIM.

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:9]: done

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:45]: done

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:16]: Thanks @ryantwolf. I have updated to reflect the first line you have. After merging this PR, Ill add this exact deduplication with GPU to future work. We do not have so much data to demo this now. The custom tokenization scripts and DAPT support did not exist in the recent containers. I think that feature just got merged. We are yet to test it. To keep things simple, we tried using one container for data curation and DAPT tutorials. Edited to remove the line.

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:38]: done

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:1]: done

File: tutorials/dapt-curation/code/main.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/main.py:63]: done

File: tutorials/dapt-curation/code/main.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/main.py:84]: done

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:52]: It was a placeholder. Removed it

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:20]: Typo. Removed it

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:21]: done

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:46]: We described this in readme. Adding the default path to the comments.

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:48]: Added path

File: tutorials/dapt-curation/code/sources/arxiv_urls.jsonl
Comment (jvamaraju) [reply tutorials/dapt-curation/code/sources/arxiv_urls.jsonl:1]: jsonl seems to be the widely used format even by chipnemo team. Lets keep it for now

File: tutorials/dapt-curation/code/main.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/main.py:212]: Users can use optional parameters to set number of workers, cpu/gpu etc right?

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:170]: To force the function call and makes it easy to visualize before next calls.

File: tutorials/dapt-curation/code/docbuilder.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/docbuilder.py:90]: Yes. It works for a few wiki pages as well

File: tutorials/dapt-curation/code/docbuilder.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/docbuilder.py:102]: done

File: tutorials/dapt-curation/code/docbuilder.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/docbuilder.py:140]: Yes Ryan. The urls need not be downloaded to htmls. We are parsing and extracting them directly. We can leave this option for the users to experiment with local files.

File: tutorials/dapt-curation/code/downloaders.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/downloaders.py:213]: removed it

File: tutorials/dapt-curation/code/docbuilder.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/docbuilder.py:215]: done

File: tutorials/dapt-curation/code/main.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/main.py:120]: Corrected it. I assumed after deduplication, filter and clean would work on lesser number of files and hence computation reduces. Makes sense.

File: tutorials/dapt-curation/code/main.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/main.py:132]: Done

File: tutorials/dapt-curation/code/main.py
Comment (Maghoumi) [reply tutorials/dapt-curation/code/main.py:120]: Yes exactly. Thanks.

File: tutorials/dapt-curation/code/utils.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/utils.py:285]: @ryantwolf I'm not sure how to leverage nemo_curator.Filter for this as it applies a filter simultaneously based on two fields. Can you point me to any documentation that can help?

**Comment by jvamaraju** [comment]
@ryantwolf I have incorporated all edits based on your feedback. Just left with filter_text_lines and filter_code_lines. Please take a look and let me know if anything else is needed.

File: tutorials/dapt-curation/code/utils.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/utils.py]: @ryantwolf @Maghoumi I have updated the code to use scorefilters now. Please review.

File: tutorials/dapt-curation/README.md
Comment (ryantwolf) [inline tutorials/dapt-curation/README.md:3]: NIMs are microservices. I don't see you using a microservice in the tutorial, so you probably are looking for a different word.

Please clarify the relationship between ChipNeMo, Llama 2, and "NeMo foundation models" in this paragraph. You introduce all of them and it's unclear what the connection is.

File: tutorials/dapt-curation/code/main.py
Comment (ryantwolf) [reply tutorials/dapt-curation/code/main.py:233]: Makes sense.

File: tutorials/dapt-curation/code/utils.py
Comment (ryantwolf) [inline tutorials/dapt-curation/code/utils.py:285]: Nit: Make class casing consistent. `LineCountFilter_text ` -> `TextLineCountFilter`.

File: .pre-commit-config.yaml
Comment (ryantwolf) [inline .pre-commit-config.yaml:1]: Undo the changes here.

File: tutorials/dapt-curation/code/utils.py
Comment (ryantwolf) [inline tutorials/dapt-curation/code/utils.py:305]: Nit: Make class casing consistent. `LineCountFilter_code ` -> `CodeLineCountFilter`.

**Review by ryantwolf — CHANGES_REQUESTED** [review]
Looks pretty good overall, thanks for cleaning it up. Just a few nits and we should be good.

File: tutorials/dapt-curation/README.md
Comment (jvamaraju) [reply tutorials/dapt-curation/README.md:3]: Refined the text. Please review

File: tutorials/dapt-curation/code/utils.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/utils.py:285]: Its already consistent right? I dont see this in other tutorials

Comment (ryantwolf) [reply tutorials/dapt-curation/code/utils.py:285]: Class names should be consistent in style with other class names in NeMo Curator and the [Python style](https://peps.python.org/pep-0008/#class-names) for class names. That means no underscores in the class name, and it should follow the CapWords convention.

File: tutorials/dapt-curation/README.md
Comment (ryantwolf) [reply tutorials/dapt-curation/README.md]: Thanks, remove the word "NIM" and we should be good

Comment (jvamaraju) [reply tutorials/dapt-curation/README.md]: Done

File: tutorials/dapt-curation/code/utils.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/utils.py:285]: Got it. Makes sense :) Corrected it

File: tutorials/dapt-curation/code/utils.py
Comment (jvamaraju) [reply tutorials/dapt-curation/code/utils.py]: Done

**Pull request merged**
